### PR TITLE
Add trace_id to log payload to enable log aggregation in GAE Flex

### PR DESCRIPTION
**Pull request by liyanhui1228**
Extract trace_id from HTTP request header and add it to application log. The pantheon UI will show the app logs grouped by trace_id and show them aggregated in one request.

Below is the screenshot of the aggregated log:
screenshot from 2017-05-01 17 37 44

### CODE REVIEW

File: logging/google/cloud/logging/handlers/_helpers.py [hunk 1]
```diff
@@ -36,4 +38,44 @@ def format_stackdriver_json(record, message):
...
+    # Extract the trace_id
+    if flask:
+        try:
+            trace_id = flask.request.headers['X-Cloud-Trace-Context'].split('/')[0]
+        except KeyError:
+            trace_id = None
+    else:
+        trace_id = None
+
+    return trace_id
```
Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: Please end your file with a newline

File: logging/google/cloud/logging/handlers/_helpers.py [hunk 2]
```diff
@@ -36,4 +38,44 @@ def format_stackdriver_json(record, message):
...
+        :return: trace_id get from flask header
+    """
+    # Import the web framework to parse HTTP header.
+    try:
+        import flask
+    except ImportError:
+        flask = None
+
+    # Extract the trace_id
+    if flask:
```
Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: `if flask is not None`

File: logging/google/cloud/logging/handlers/_helpers.py [hunk 3]
```diff
@@ -36,4 +38,44 @@ def format_stackdriver_json(record, message):
...
+    """Parse flask request header to get trace_id.
+    
+        :rtype: str
+        :return: trace_id get from flask header
+    """
+    # Import the web framework to parse HTTP header.
+    try:
+        import flask
+    except ImportError:
+        flask = None
```
Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: Please do this at module-scope

File: logging/google/cloud/logging/handlers/_helpers.py [hunk 4]
```diff
@@ -36,4 +38,44 @@ def format_stackdriver_json(record, message):
...
+        payload['traceId'] = trace_id
+
     return json.dumps(payload)
+
+
+def parse_flask_request_header():
+    """Parse flask request header to get trace_id.
+    
+        :rtype: str
+        :return: trace_id get from flask header
```
Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: Should de-dent these to be at the same indent level as everything else

File: logging/google/cloud/logging/handlers/_helpers.py [hunk 5]
```diff
@@ -36,4 +38,44 @@ def format_stackdriver_json(record, message):
         'severity': record.levelname,
     }
 
+    # Make a best effort to add the trace id.
+    # First try to extract the trace_id from HTTP header, only support flask framework for now.
+    trace_id = parse_flask_request_header()
```
Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: I am big time :-1: on putting a **flask-specific** check in a generic method.

A better fix would probably involve allowing a way to register extra checkers, either through a registry in this module or through a registry on a `Client()` object (or a similar type).

**Comment by waprin** [comment]
This whole PR should get closed, we are just supposed to use the generic log handler according to @justinbeckwith in #2997

**Comment by waprin** [comment]
Maybe we can still get the app ID in there but it shouldn't be via the stackdriver helper that currently exists.

**Comment by dhermes** [comment]
@waprin Can we make a decision quickly and either move forward or close this thing? (i.e. just don't want it to come off our plates and sit open in limbo)

**Comment by waprin** [comment]
I think we should close and @liyanhui1228 and @duggelz should have a quick chat about best way to do this.

File: logging/google/cloud/logging/handlers/_helpers.py (hunk 5 above)
Comment (duggelz) [inline logging/google/cloud/logging/handlers/_helpers.py]: Adding Flask-specific, Django-specific, webapp2-specific, pycharm-specific, etc code to support the silver language contract has always been my understanding of the plan.  The exact implementation undetermined, and so now is a good time to talk about it.  Do you have access to the "Stackdriver: Silver Language Requirements" document?  Specifically the "Runtimes: Logging" section.  Shall I setup a meeting to discuss, or take it to email?

Comment (dhermes) [inline logging/google/cloud/logging/handlers/_helpers.py]: @duggelz The issue is not **THAT** we provide purpose-built helpers, just **WHERE / HOW**. (Hence my suggestion for a better fix.)

**Comment by dhermes** [comment]
OK, closing now. We can always re-open.

**Comment by duggelz** [comment]
This has nothing to do with #2997

**Comment by jonparrott** [comment]
Okay - deep breath - we failed to properly support a new contributor here. We need to do better.

It's clear that while we *want* this feature, we're not ready to implement it. So, let's create a bug/feature request for this feature and mark all blocking bugs. If the owners of those bugs can't resolve them in a timely manner, let's see what we can do to enable @liyanhui1228 or @duggelz to address them so that this feature can be implemented.

@dhermes or @lukesneeringer, can you own this? If not, please let me know and I will do this personally.

**Comment by lukesneeringer** [comment]
I am happy to own it if needed ("buck stops here" kind of thing), but I am not sure I am the right person. I have read over this ticket and I do not intuitively grok the issue in question.

I _think_ the issue is that if there is an HTTP header that we can easily get at, we want to send it to Stackdriver in some way?

Best case for me is if @jonparrott owns this, but I am willing to if needed. :-)

**Comment by jonparrott** [comment]
@lukesneeringer I'm on it.

**Pull request closed without merging**